File: java/examples/src/main/java/com/manycoreapis/sdk/examples/WorldReconstruct.java
package com.manycoreapis.sdk.examples;

import com.manycoreapis.sdk.asset.AssetClient;
import com.manycoreapis.sdk.asset.UploadResult;
import com.manycoreapis.sdk.core.AholoClientConfig;
import com.manycoreapis.sdk.world.WorldClient;
import com.manycoreapis.sdk.world.model.ReconstructionCreateParams;
import com.manycoreapis.sdk.world.model.WorldAsyncOperation;
import com.manycoreapis.sdk.world.model.WorldDetail;

import java.nio.file.Paths;

/**
 * Minimal example: upload a video, create a 3DGS reconstruction, poll until done.
 *
 * Usage:
 *   AHOLO_API_KEY=xxx mvn exec:java -pl examples \
 *     -Dexec.mainClass=com.manycoreapis.sdk.examples.WorldReconstruct \
 *     -Dexec.args="/path/to/room.mp4"
 */
public class WorldReconstruct {
    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.err.println("Usage: WorldReconstruct <video-path>");
            System.exit(1);
        }

        String filePath = args[0];
        String region = System.getenv().getOrDefault("AHOLO_REGION", "cn");
        AholoClientConfig config = AholoClientConfig.ofRegion(region);
        AssetClient asset = AssetClient.create(config);
        WorldClient world = WorldClient.create(config);

        System.out.println("Uploading " + filePath + " ...");
        long t0 = System.currentTimeMillis();

        UploadResult uploaded = asset.uploadFile(Paths.get(filePath), (uploadedBytes, total) -> {
            int pct = (int) (uploadedBytes * 100 / total);
            System.out.printf("\r  %d%% (%.1f/%.1f MB)", pct,
                    uploadedBytes / 1024.0 / 1024.0, total / 1024.0 / 1024.0);
            System.out.flush();
        });
        System.out.println();
        System.out.printf("Upload complete (%dms) url=%s%n", System.currentTimeMillis() - t0, uploaded.url());

        System.out.println("Creating reconstruction task...");
        WorldAsyncOperation created = world.reconstructions().create(
                ReconstructionCreateParams.builder()
                        .name("SDK reconstruction demo")
                        .addResource(uploaded.url(), "video")
                        .taskQuality("normal")
                        .scene("model")
                        .build()
        );
        String worldId = created.worldId();
        System.out.println("worldId=" + worldId + ", polling...");

        t0 = System.currentTimeMillis();
        WorldDetail detail = world.waitFor(worldId);
        System.out.printf("Reconstruction complete (%dms)%n", System.currentTimeMillis() - t0);
        System.out.println(detail);
        detail.assets().ifPresent(assets ->
                assets.splats().ifPresent(splats ->
                        splats.urls().ifPresent(urls -> {
                            urls.plyPath().ifPresent(p -> System.out.println("PLY: " + p));
                            urls.spzPath().ifPresent(p -> System.out.println("SPZ: " + p));
                        })
                )
        );
    }
}

File: python/examples/world_reconstruct.py
#!/usr/bin/env python3
"""Upload media and run a 3DGS reconstruction.

Supported inputs: `.mp4` / `.mov` (type `video`), Insta360 `.insv` (type `insv`).

Usage:
  AHOLO_API_KEY=xxx python world_reconstruct.py ./room.mp4
  AHOLO_API_KEY=xxx python world_reconstruct.py ./panorama.insv
"""
import json
import os
import sys
import time

from manycore.aholo_sdk_asset import create_asset_client
from manycore.aholo_sdk_core import AholoClientConfig
from manycore.aholo_sdk_world import create_world_client


def reconstruction_resource_type(file_path: str) -> str:
    ext = os.path.splitext(file_path)[1].lower()
    if ext == ".insv":
        return "insv"
    if ext in {".mp4", ".mov"}:
        return "video"
    raise SystemExit(f'Unsupported extension "{ext}". Use .mp4, .mov, or .insv')


if len(sys.argv) < 2:
    raise SystemExit("Usage: AHOLO_API_KEY=xxx python world_reconstruct.py <media-path>")

file_path = sys.argv[1]
resource_type = reconstruction_resource_type(file_path)
region = os.environ.get("AHOLO_REGION", "cn")
config = AholoClientConfig(region=region)
asset = create_asset_client(config)
world = create_world_client(config)


def show_progress(uploaded_bytes: int, total: int) -> None:
    pct = int(uploaded_bytes / total * 100)
    print(f"\r  {pct}% ({uploaded_bytes / 1024 / 1024:.1f}/{total / 1024 / 1024:.1f} MB)", end="", flush=True)


print(f"Uploading {file_path} (type={resource_type}) ...")
t0 = time.monotonic()
uploaded = asset.upload_file(file_path, on_progress=show_progress)
print()
print(f"Upload complete ({int((time.monotonic() - t0) * 1000)}ms) url={uploaded.url}")

print("Creating reconstruction task...")
op = world.reconstructions.create(
    resources=[{"url": uploaded.url, "type": resource_type}],
    task_quality="normal",
    scene="model",
    name="SDK reconstruction demo",
)
world_id = op["worldId"]
print(f"worldId={world_id}, polling...")

t0 = time.monotonic()
detail = world.wait_for(world_id)
print(f"Reconstruction complete ({int((time.monotonic() - t0) * 1000)}ms)")
print(json.dumps(detail, indent=2))
